### DISCUSSION TITLE
is it okay to use imgui for iOS apps?

### DISCUSSION THREAD

**Discussion by rahil627** [ocornut/imgui#5090] [New Users Build/Link/Run issues ONLY!]
so, i’m just researching here, but i’ve read quite a few times that the main problem with immediate gui, in general, as opposed to the normal “retained” gui, is that it can be CPU-expensive, for drawing the entire GUI on every frame.

First, is that even true? lol

I personally just want to make simple apps for iOS, but i am not so interested in any of the current pop solutions (C#’s Avalonia/Uno/MAUI nor Google’s Flutter), and i don’t care about native *feel* anyway. Although, i do hope it’s not too bad to bind some iOS api(??), like the share button. I come from the 2d game-making world, and imgui just seems like a *much much* simpler solution for me, in addition to being an awesome tool for programming in general!

For the sake of simplicity, let’s use a text editor with a Dropbox file-browser.

Will it really drain the battery? Are the native iOS text-box widget and tree widget (if that exists) really that complex?

Or is this all a pipe dream? :(

anyway, hope to use it soon for general purposes anyway :)

(huh, sorry to post here, as i see the only category is issues... but i read on the actual forums that it was better to post here..)

**Comment by PathogenDavid** [comment]
TL;DR: Just because you can doesn't mean you should.

> is it okay to use imgui for iOS apps?

There's no reason you couldn't (and people do), but I personally would not unless it's just for debugging stuff. IMO using Dear ImGui as the primary UI for mobile apps is shoving a square peg in the round hole. It was designed around desktop UI paradigms, not mobile ones.

>  i’ve read quite a few times that the main problem with immediate gui, in general, as opposed to the normal “retained” gui, is that it can be CPU-expensive, for drawing the entire GUI on every frame.
>
> First, is that even true? lol

I wouldn't say this is inherently true, but it is somewhat true for Dear ImGui. Dear ImGui was first and foremost designed for games, in which situation you're probably rendering a lot more complex stuff at 60 FPS regardless so its overhead if negligible.

There are some workarounds for this, look at the various power saving PRs. https://github.com/ocornut/imgui/pull/4076 https://github.com/ocornut/imgui/pull/3124 https://github.com/ocornut/imgui/pull/2749

> I personally just want to make simple apps for iOS, [...], and i don’t care about native feel anyway.

I would be concerned about cert. Dear ImGui won't play nice with VoiceOver and I don't believe the text input would properly support things like auto complete or correction. (If it does I can't imagine it works as well as the native text input widgets.) I don't have experience with cert on iOS but my understanding is Apple is very picky and I can't imagine they'd like these things.

Also my understanding is Apple rejects apps which they think depart too far from their UI guidelines.

> (huh, sorry to post here, as i see the only category is issues... but i read on the actual forums that it was better to post here..)

Not sure what "actual forums" you're referring to, but an issue would've been preferred.

**Reply by stuartcarnie** [reply]
> I would be concerned about cert. Dear ImGui won't play nice with VoiceOver and I don't believe the text input would properly support things like auto complete or correction. (If it does I can't imagine it works as well as the native text input widgets.) I don't have experience with cert on iOS but my understanding is **Apple is very picky and I can't imagine they'd like these things**.

Whilst I think your point is valid regarding VoiceOver and accessibility, stating that 

> Apple is very picky and I can't imagine they'd like these things

is verifiably not true, as you only need to look at all the games on the App Store, which almost exclusively provide their own custom UI that often don't play nice with these features.

Source: I have previously released C64 for iPhone and several other games on the App store

**Comment by rahil627** [comment]
*wow*, thanks so much for replying so thoroughly :heart:

the forum thread i saw was this one, as it had a person who makes iOS music apps thinking about porting over to imgui:
https://discourse.dearimgui.org/t/imgui-for-a-semi-pro-audio-app/337

and ah yes, now thinking about the text editor example even more, the text widget actually does have an insane amount of work put into it, perhaps one of the most complex widgets! auto-complete, spelling, voice, emoji, this, that... native (swift) feels like the only choice there...

ahhh, i didn’t think about Apple’s standards... bummer. Well, *wow...* that... that really narrows down my solutions... Even if it wasn’t an immediate gui, and i just made an app using some non-native ui library, it could still be a problem? :sob:

dang, that really kills that idea then. I was hoping to just use a game engine i’m familiar with + ui to make iOS apps... instead of, you know, learning and using a whole ‘nother framework and language... All of life should be like making games! 😜
